logger_test.go

```diff
@@ -124,20 +125,104 @@ func TestLoggerInitialFields(t *testing.T) {
-		// Child loggers should have copy-on-write semantics, so two children
-		// shouldn't stomp on each other's fields or affect the parent's fields.
-		logger.With(String("one", "two")).Info("")
-		logger.With(String("three", "four")).Info("")
-		logger.Info("")
+	for _, tt := range []struct {
+		name string
+		opts []Option
+	}{
+		{
+			"regular non deferred logger",
+			[]Option{Fields(Int("foo", 42))},
+		},
+		{
+			"deferred logger",
+			[]Option{Fields(Int("foo", 42)), DeferWith(true)},
+		},
+	} {
+		t.Run(tt.name, func(t *testing.T) {
+			withLogger(t, DebugLevel, tt.opts, func(logger *Logger, logs *observer.ObservedLogs) {
+				// Child loggers should have copy-on-write semantics, so two children
+				// shouldn't stomp on each other's fields or affect the parent's fields.
+				logger.With(String("one", "two")).Info("")
+				logger.With(String("three", "four")).Info("")
+				logger.With(String("five", "six")).With(String("seven", "eight")).Info("")
+				logger.Info("")
+
+				assert.Equal(t, []observer.LoggedEntry{
+					{Context: []Field{Int("foo", 42), String("one", "two")}},
+					{Context: []Field{Int("foo", 42), String("three", "four")}},
+					{Context: []Field{Int("foo", 42), String("five", "six"), String("seven", "eight")}},
+					{Context: []Field{Int("foo", 42)}},
+				}, logs.AllUntimed(), "Unexpected cross-talk between child loggers.")
+			})
+		})
+	}
+}
+
+func TestLoggerDeferredWithCapture(t *testing.T) {
+
+	type res struct {
+		A, B, C, D, E []int
+	}
+	for _, tt := range []struct {
+		name              string
+		opts              []Option
+		wantRes, wantRes2 res
+	}{
+		{
+			name:     "non deferred logger evaluates at time of With()",
+			wantRes:  res{A: []int{0}, B: []int{1}, C: []int{2}},
+			wantRes2: res{A: []int{0}, B: []int{1}, C: []int{2}, D: []int{3}, E: []int{4}},
+		},
+		{
+			name:     "non deferred logger evaluates at time of With(), with deferWith explicit false",
+			opts:     []Option{DeferWith(false)},
+			wantRes:  res{A: []int{0}, B: []int{1}, C: []int{2}},
+			wantRes2: res{A: []int{0}, B: []int{1}, C: []int{2}, D: []int{3}, E: []int{4}},
+		},
+		{
+			name:     "deferred logger evaluates at time of logging, and stops deferring on children",
+			opts:     []Option{DeferWith(true)},
+			wantRes:  res{A: []int{2}, B: []int{2}, C: []int{2}},
+			wantRes2: res{A: []int{2}, B: []int{2}, C: []int{2}, D: []int{3}, E: []int{4}},
+		},
+	} {
+		t.Run(tt.name, func(t *testing.T) {
+			bs := &ztest.Buffer{}
+			enc := zapcore.NewJSONEncoder(NewProductionConfig().EncoderConfig)
+			core := zapcore.NewCore(enc, bs, DebugLevel)
+			logger := New(core, tt.opts...)
+
+			var x int
+			arr := zapcore.ArrayMarshalerFunc(func(enc zapcore.ArrayEncoder) error {
+				enc.AppendInt(x)
+				return nil
+			})
+
+			logger = logger.With(Array("a", arr))
+			x = 1
+			logger = logger.With(Array("b", arr))
+			x = 2
+			logger.Info("hello", Array("c", arr))
+			x = 3
```

Comment:
@rabbbit @abhinav do you agree this is the correct behaviour?

Reply:
I think so, but the example seems painful. I don't know `zap` enough to figure out an easier one (@abhinav?). It does seem like "a" and "b" are always equal - could you simplify and remove if they're really unnecessarily duplicated?

And generally, I'd personally split the PR into two or even four:
- one that modifies `TestLoggerWith` - this seems non-controversial, perhaps adding benchmarks too.
- one that adds `TestLoggerDeferredCapture` to formalize the current behavior as a test
- one that adds the `WithDefer` behavior.

Apologies for being a pain.

Reply:
1a/ https://github.com/uber-go/zap/pull/1325 (benchmarks)
2/ https://github.com/uber-go/zap/pull/1326 (deferred capture - cleaner)